// @vkid/captcha/dist/esm/utils/checkCaptchaError/checkCaptchaError.js
import { Widget } from '../../core/widget/widget.js';
import { CheckCaptchaType } from './types.js';
import { getChallengeUrlDomain } from './utils.js';

function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
var ERROR_CODE_CAPTCHA = 14;
var ERROR_TYPE_CAPTCHA = 'captcha';
var checkCaptchaError = function(param) {
    var responseHeaders = param.responseHeaders, url = param.url, responseError = param.responseError, withWidget = param.withWidget, widgetConfig = param.widgetConfig;
    var isChallengeRequired = responseHeaders.get('X-Challenge') === 'required';
    var challengeUrl = responseHeaders.get('X-Challenge-Url');
    var traceId = responseHeaders.get('X-Trace-Id');
    var isCaptchaType2 = isChallengeRequired && !!challengeUrl;
    var isApiCaptchaError = !!responseError && 'error_code' in responseError && responseError.error_code === ERROR_CODE_CAPTCHA;
    var isLoginCaptchaError = !!responseError && 'type' in responseError && responseError.type === ERROR_TYPE_CAPTCHA;
    var isCaptchaType1 = (isApiCaptchaError || isLoginCaptchaError) && !!responseError.redirect_uri;
    var isUnknownCaptcha = (isApiCaptchaError || isLoginCaptchaError) && !responseError.redirect_uri;
    if (isCaptchaType2) {
        var iframeSrc = "https://".concat(getChallengeUrlDomain(url)).concat(challengeUrl);
        if (!withWidget) {
            return {
                captchaType: CheckCaptchaType.TYPE_2,
                captchaUrl: iframeSrc
            };
        }
        var widget = new Widget(widgetConfig);
        return {
            captchaType: CheckCaptchaType.TYPE_2,
            captchaUrl: iframeSrc,
            captchaWidget: {
                show: function(params) {
                    return widget.show(_object_spread_props(_object_spread({}, params), {
                        iframeSrc: iframeSrc,
                        traceId: traceId,
                        captchaType: CheckCaptchaType.TYPE_2,
                        requestUrl: url
                    }));
                },
                close: function() {
                    return widget.close();
                },
                update: function(config) {
                    return widget.update(config);
                }
            }
        };
    }
    if (isCaptchaType1 && !!responseError.redirect_uri) {
        var iframeSrc1 = responseError.redirect_uri;
        if (!withWidget) {
            return {
                captchaType: CheckCaptchaType.TYPE_1,
                captchaUrl: iframeSrc1
            };
        }
        var widget1 = new Widget(widgetConfig);
        return {
            captchaType: CheckCaptchaType.TYPE_1,
            captchaUrl: iframeSrc1,
            captchaWidget: {
                show: function(params) {
                    return widget1.show(_object_spread_props(_object_spread({}, params), {
                        iframeSrc: iframeSrc1,
                        captchaType: CheckCaptchaType.TYPE_1
                    }));
                },
                close: function() {
                    return widget1.close();
                },
                update: function(config) {
                    return widget1.update(config);
                }
            }
        };
    }
    if (isUnknownCaptcha) {
        return {
            captchaType: CheckCaptchaType.UNKNOWN
        };
    }
    return {
        captchaType: null
    };
};

export { checkCaptchaError };
